### PR TITLE
Fixing typos, allowing override of nil default properties, functional override modules

### PR DESCRIPTION
**Pull request by Nimblz**
This pull fixes issues #13 #12 #11 

Let me know if I made any mistakes!

### CODE REVIEW

File: src/defineComponent.lua [hunk 1]
```diff
@@ -29,16 +28,13 @@ local function defineComponent(tagName, defaultProps)
     definition.tagName = tagName
 
     function definition._create(instance)
-        local component
-        if definition.defaultProps then
```
Comment (AmaranthineCodices) [inline src/defineComponent.lua]: This change will break code that does not specify `defaultProps` via the second argument:

```lua
local SomeComponent = defineComponent("Something")

-- This function will never be called.
function SomeComponent.defaultProps()
    return {
        a = 1,
    }
end
```

File: src/defineComponent.lua [hunk 2]
```diff
@@ -9,11 +9,10 @@ local createCleaner = require(script.Parent.createCleaner)
 local errorFormats = {
     nonStringName = "tagName (1) must be a string, is a %s",
     nonFunctionDefaultProps = "defaultProps (2) must be a function or nil, is a %s",
-    nonTableDefaultPropsReturn = "The defaultProps generator for the %s component must return a table, but returned a %s",
+    nonTableDefaultPropsReturn = "The defaultProps generator for the %s component must return a function, but returned a %s",
```
Comment (AmaranthineCodices) [inline src/defineComponent.lua]: The return value from the default props generator function must be a table, not a function - this change is incorrect.

File: src/defineComponent.lua [hunk 3]
```diff
@@ -52,20 +48,17 @@ local function defineComponent(tagName, defaultProps)
 
             local componentProps = require(componentPropsModule)
```
Comment (AmaranthineCodices) [inline src/defineComponent.lua]: For clarity, can we instead call this `componentPropsGenerator`, then use `componentProps` as the return value of this?

File: src/defineComponent.lua [hunk 4]
```diff
@@ -29,16 +28,13 @@ local function defineComponent(tagName, defaultProps)
...
+        local component = {}
+        if defaultProps then
+            component = defaultProps()
             assert(
-                typeof(component) == "table",
-                errorFormats.nonTableDefaultPropsReturn:format(tagName, typeof(component)))
-        else
-            component = {}
+                type(component) == "table",
+                errorFormats.nonTableDefaultPropsReturn:format(tagName, type(component)))
```
Comment (AmaranthineCodices) [inline src/defineComponent.lua]: Use `typeof` here, for better information about Roblox-specific types.

File: src/defineComponent.lua [hunk 5]
```diff
@@ -52,20 +48,17 @@ local function defineComponent(tagName, defaultProps)
 
             local componentProps = require(componentPropsModule)
             assert(
-                typeof(componentProps) == "table",
-                errorFormats.nonTableRootPropOverride:format(fullName, typeof(componentProps)))
+                typeof(componentProps) == "function",
+                errorFormats.nonFunctionRootPropOverride:format(fullName, typeof(componentProps)))
 
-            local specificComponentProps = componentProps[tagName]
+            local componentPropsOverride = componentProps()
```
Comment (AmaranthineCodices) [inline src/defineComponent.lua]: Rename to `componentProps`

File: src/defineComponent.lua (hunk 4 above)
Comment (Nimblz) [reply src/defineComponent.lua]: didn't know typeof worked for the lua types, thought it was just for roblox userdata

**Pull request merged**